File: dagaar_catering/catering_management/controllers/catering_production_plan.py
# Copyright (c) 2024, DagaarSoft — catering_production_plan.py

import frappe
from frappe import _
from frappe.utils import flt


def validate(doc, method=None):
	_fetch_order_data(doc)
	if not doc.items:
		frappe.throw(_("Production Plan must have at least one item."))


def on_submit(doc, method=None):
	frappe.db.set_value("Catering Order", doc.catering_order, {
		"production_plan": doc.name,
		"status": "In Production"
	})


def _fetch_order_data(doc):
	if doc.catering_order:
		order = frappe.db.get_value(
			"Catering Order", doc.catering_order,
			["customer_name", "event_date", "total_guests", "company"], as_dict=True
		)
		if order:
			doc.customer = order.customer_name
			doc.event_date = order.event_date
			doc.total_guests = order.total_guests
			if not doc.company:
				doc.company = order.company

File: dagaar_catering/public/js/catering_production_plan.js
// catering_production_plan.js
frappe.ui.form.on("Catering Production Plan", {
	refresh(frm) {
		if (frm.doc.docstatus === 1) {
			// Bulk-create Work Orders from production plan items
			frm.add_custom_button(__("Create Work Orders"), () => {
				frappe.call({
					method: "dagaar_catering.catering_management.controllers.catering_order.create_work_orders_from_plan",
					args: { production_plan: frm.doc.name },
					freeze: true,
					freeze_message: __("Creating Work Orders..."),
					callback: (r) => {
						if (r.message && r.message.length) {
							frappe.show_alert({
								message: __("Created {0} Work Orders", [r.message.length]),
								indicator: "green"
							}, 6);
							// Open each Work Order in a new tab
							r.message.forEach(wo => {
								window.open(`/app/work-order/${encodeURIComponent(wo)}`, "_blank");
							});
							frm.reload_doc();
						}
					}
				});
			});
		}
	},
});

frappe.ui.form.on("Catering Production Plan Item", {
	work_order(frm, cdt, cdn) {
		const row = locals[cdt][cdn];
		if (row.work_order) {
			// When clicking the work order link, open in new tab
			const url = `/app/work-order/${encodeURIComponent(row.work_order)}`;
			window.open(url, "_blank");
		}
	},
});
